Initial checkin of first tracing notebook

tgasser-nv opened the pull request:
## Description

<!-- Describe the big picture of your changes to communicate to the maintainers
  why we should accept this pull request. -->

## Related Issue(s)

<!--
  If this PR fixes any issues, please link to the issue here.
  - Fixes #<issue_number>
-->

## Checklist

- [ ] I've read the [CONTRIBUTING](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [ ] I've updated the documentation if applicable.
- [ ] I've added tests if applicable.
- [ ] @mentions of the person or team responsible for reviewing proposed changes.

Pouyanpi reviewed (changes requested):
Thank you @tgasser-nv. This is a valuable addition 👍🏻 


**Merge plan**: Let's ensure this is merged once all pending PRs for 0.16.0 are merged to `develop` and do a rebase on top of this so that it reflects the latest changes.

**Docs build**: @miyoungc can you confirm that build_notebook_docs.py is the right path for rendering this notebooks to
https://docs.nvidia.com/nemo/guardrails/latest/getting-started ?

**Tutorial usage**: right now users can’t follow the docs due to missing configs. Two options, happy with either here:
* Inline the needed configs in the notebook and instruct users to create a config/ folder (like other tutorials), or
* Use `from_content` to avoid file creation entirely.

**Nice to haves** (follow-ups, separate PR):
* Create NemoGuard NIMs configs to examples/configs so cloned users can reference them directly.
* Use the most recent `span_format: opentelemetry`. (I’ll switch the default in #1331; this PR can keep current unless trivial.) The current default format is considered legacy and we might deprecate it.
* Show how to toggle parallel/sequential dynamically so we don’t need two folders:

```
config.rails.input.parallel = True
config.rails.output.parallel = False

```

* point users to the “Parallel Execution of Input and Output Rails” section.
docs: https://docs.nvidia.com/nemo/guardrails/latest/user-guides/configuration-guide/guardrails-configuration.html#parallel-rails

* quiet installs to keep notebook output clean:
```
!pip install pandas plotly langchain_nvidia_ai_endpoints aiofiles -q
```

If we pick one of the two config approaches above and rebase after 0.16.0 lands it'd be great.

cparisien commented on docs/getting-started/8-tracing/1_tracing_quickstart.ipynb:
```diff
@@ -0,0 +1,4713 @@
+   "source": [
+    "import yaml\n",
+    "\n",
+    "def load_config_trace_filename(config_filename):\n",
+    "    \"\"\"Helper function to load FileSystem tracing filename\"\"\"\n",
+    "\n",
```
Oh we're just removing the previous trace files if they exist? That's fine, just clarify that's all this is for.

cparisien commented on docs/getting-started/8-tracing/1_tracing_quickstart.ipynb:
```diff
@@ -0,0 +1,4713 @@
+       "    </tr>\n",
+       "  </thead>\n",
+       "  <tbody>\n",
+       "    <tr>\n",
+       "      <th>0</th>\n",
+       "      <td>interaction</td>\n",
+       "      <td>8.600522</td>\n",
+       "    </tr>\n",
```
This one looks like the overall, end-to-end latency? i.e., it's the top-tier and the rest in this table are a level below this one

Pouyanpi commented on docs/getting-started/8-tracing/1_tracing_quickstart.ipynb:
```diff
@@ -0,0 +1,4713 @@
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "ERROR:nemoguardrails.rails.llm.llmrails:Failed to create isolated LLM instance for action 'self_check_hallucination'. This is required to prevent parameter contamination between different actions. \n",
+      "\n",
+      "Possible solutions:\n",
+      "1. If using a custom LLM class, ensure it supports copy.copy() operation\n",
+      "2. Check that your LLM configuration doesn't contain non-copyable objects\n",
+      "3. Consider using a dedicated LLM configuration for action 'self_check_hallucination'\n",
+      "\n",
+      "Original error: \"ChatNVIDIA\" object has no field \"model_kwargs\"\n",
+      "\n",
+      "To use a dedicated LLM for this action, add to your config:\n",
+      "models:\n",
+      "  - type: self_check_hallucination\n",
+      "    engine: <your_engine>\n",
+      "    model: <your_model>\n",
+      "WARNING:nemoguardrails.rails.llm.llmrails:Failed to create isolated LLMs for actions: Failed to create isolated LLM instance for action 'self_check_hallucination'. This is required to prevent parameter contamination between different actions. \n",
+      "\n",
+      "Possible solutions:\n",
+      "1. If using a custom LLM class, ensure it supports copy.copy() operation\n",
+      "2. Check that your LLM configuration doesn't contain non-copyable objects\n",
+      "3. Consider using a dedicated LLM configuration for action 'self_check_hallucination'\n",
+      "\n",
+      "Original error: \"ChatNVIDIA\" object has no field \"model_kwargs\"\n",
+      "\n",
+      "To use a dedicated LLM for this action, add to your config:\n",
+      "models:\n",
+      "  - type: self_check_hallucination\n",
+      "    engine: <your_engine>\n",
+      "    model: <your_model>\n"
```
this is resolved by #1336

tgasser-nv replied:
Yes just rebased on top of Pouyan's PR and the ERRORs are gone

tgasser-nv commented:
> Thank you @tgasser-nv. This is a valuable addition 👍🏻
> 
> **Merge plan**: Let's ensure this is merged once all pending PRs for 0.16.0 are merged to `develop` and do a rebase on top of this so that it reflects the latest changes.
> 
> **Docs build**: @miyoungc can you confirm that build_notebook_docs.py is the right path for rendering this notebooks to https://docs.nvidia.com/nemo/guardrails/latest/getting-started ?
> 
> **Tutorial usage**: right now users can’t follow the docs due to missing configs. Two options, happy with either here:
> 
> * Inline the needed configs in the notebook and instruct users to create a config/ folder (like other tutorials), or
> * Use `from_content` to avoid file creation entirely.
> 
> **Nice to haves** (follow-ups, separate PR):
> 
> * Create NemoGuard NIMs configs to examples/configs so cloned users can reference them directly.
> * Use the most recent `span_format: opentelemetry`. (I’ll switch the default in #1331; this PR can keep current unless trivial.) The current default format is considered legacy and we might deprecate it.
> * Show how to toggle parallel/sequential dynamically so we don’t need two folders:
> 
> ```
> config.rails.input.parallel = True
> config.rails.output.parallel = False
> ```
> 
> * point users to the “Parallel Execution of Input and Output Rails” section.
>   docs: https://docs.nvidia.com/nemo/guardrails/latest/user-guides/configuration-guide/guardrails-configuration.html#parallel-rails
> * quiet installs to keep notebook output clean:
> 
> ```
> !pip install pandas plotly langchain_nvidia_ai_endpoints aiofiles -q
> ```
> 
> If we pick one of the two config approaches above and rebase after 0.16.0 lands it'd be great.

Quick update:
* All the PRs for v0.16 were merged today, I rebased on top of those and did a clean run of the notebook. I can see all the spurious ERROR messages are gone, and the names of rails follow the new `guardrails.request` format
* I rewrote all the configs and created them as Python dicts before converting to RailsConfig objects using Pydantic's `model_validate()` function. The prompts are now loaded from elsewhere in the repo instead of copy-pasting them into new files. The only file in the PR is the notebook.
* I'm picking up the most recent span format, I updated the `clean_trace_dataframe()` method to match.
* Added `-q` to the pip install

I met with @miyoungc who's looking at the best way to link the notebook from the rest of the docs. Am holding off merging until Monday to give her time to work on this.

miyoungc commented on docs/user-guides/tracing/adapter-configurations.md:
````diff
@@ -26,6 +26,8 @@ tracing:
       filepath: "./logs/traces.jsonl"
 ```
 
+For working examples, refer to the [Tracing Quickstart](../../getting-started/8-tracing/1_tracing_quickstart.ipynb) notebook.
````
```suggestion
For working examples, refer to the [Tracing Quickstart](https://github.com/NVIDIA/NeMo-Guardrails/tree/develop/docs/getting-started/8-tracing/1_tracing_quickstart.ipynb) notebook.
```

miyoungc commented on docs/user-guides/tracing/index.md:
```diff
@@ -19,6 +19,10 @@ With tracing, you can:
 
+## Jupyter Notebooks
+
+- [Tracing Quickstart](../getting-started/8-tracing/1_tracing_quickstart.ipynb) - A quickstart guide to tracing Guardrails requests in sequential and parallel modes.
```
```suggestion
- [Tracing Quickstart](https://github.com/NVIDIA/NeMo-Guardrails/tree/develop/docs/getting-started/8-tracing/1_tracing_quickstart.ipynb) - A quickstart guide to tracing Guardrails requests in sequential and parallel modes.
```

Pouyanpi commented on docs/getting-started/8-tracing/1_tracing_quickstart.ipynb:
a mintor type hint issue at cell 7:

```python
   CONFIG_MODELS: Dict[str, str]
```

cell 29:

```python
parallel_df.loc[sequential_df["is_rail"]],
```
is it intended (parallel_df and sequential_df mixed use)?

also

```python
parallel_input_rail_time = sequential_df.loc[
       parallel_df["name"].isin(INPUT_RAIL_NAMES), "duration"
   ].max()
```
